Sources/EpoxyCollectionView/CollectionView/Internal/CollectionViewScrollToItemHelper.swift

```diff
@@ -333,70 +318,77 @@ final class CollectionViewScrollToItemHelper {
+      guard !insetBounds.contains(itemFrame) else { return nil }
+      return itemFrame.midX < insetBounds.midX ? .left : .right
     default:
       EpoxyLogger.shared.assertionFailure("Unsupported scroll position.")
       return nil
     }
   }
 
-  // Returns the correct resting position for a scroll-to-item action for the current viewport.
-  // This will be used to determine how much farther we need to programmatically scroll on each
-  // animation tick.
-  private func targetOriginForVisibleItem(
+  /// Returns the correct content offset for a scroll-to-item action for the current viewport.
+  ///
+  /// This will be used to determine how much farther we need to programmatically scroll on each
+  /// animation tick.
+  private func targetContentOffsetForVisibleItem(
     withFrame itemFrame: CGRect,
-    inVisibleBounds visibleBounds: CGRect,
+    inBounds bounds: CGRect,
+    contentSize: CGSize,
+    adjustedContentInset: UIEdgeInsets,
     targetScrollPosition: UICollectionView.ScrollPosition,
     scrollAxis: ScrollAxis)
     -> CGPoint
   {
-    let itemSize: CGFloat
-    let contentOffset: CGFloat
-    let viewportSize: CGFloat
+    let itemPosition, itemSize, viewportSize, minContentOffset, maxContentOffset: CGFloat
+    let visibleBounds = bounds.inset(by: adjustedContentInset)
     switch scrollAxis {
     case .vertical:
+      itemPosition = itemFrame.minY
       itemSize = itemFrame.height
-      contentOffset = visibleBounds.minY
       viewportSize = visibleBounds.height
+      minContentOffset = -adjustedContentInset.top
+      maxContentOffset = -adjustedContentInset.top + contentSize.height - visibleBounds.height
     case .horizontal:
+      itemPosition = itemFrame.minX
       itemSize = itemFrame.width
-      contentOffset = visibleBounds.minX
       viewportSize = visibleBounds.width
+      minContentOffset = -adjustedContentInset.left
+      maxContentOffset = -adjustedContentInset.left + contentSize.width - visibleBounds.width
     }
 
     let newOffset: CGFloat
     switch targetScrollPosition {
     case .top, .left:
-      newOffset = contentOffset
+      newOffset = itemPosition + minContentOffset
     case .bottom, .right:
-      newOffset = contentOffset + viewportSize - itemSize
+      newOffset = itemPosition + itemSize - viewportSize + minContentOffset
     case .centeredVertically, .centeredHorizontally:
-      newOffset = contentOffset + (viewportSize / 2) - (itemSize / 2)
-    case []:
-      newOffset = 0
+      newOffset = itemPosition + (itemSize / 2) - (viewportSize / 2) + minContentOffset
     default:
       EpoxyLogger.shared.assertionFailure("Unsupported scroll position.")
       return itemFrame.origin
     }
 
-    switch scrollAxis {
-    case .vertical: return CGPoint(x: itemFrame.origin.x, y: newOffset)
-    case .horizontal: return CGPoint(x: newOffset, y: itemFrame.origin.y)
-    }
+    let clampedOffset = min(max(newOffset, minContentOffset), maxContentOffset)
```

Comment:
This ensures that we don't attempt to scroll to a position that would be outside of the valid range of content offsets, which could cause strange snapping animations where the scroll view is scrolled out of its content bounds and then snaps back in the final frame.